crafting.py

```diff
@@ -0,0 +1,316 @@
+    
+    
+    state  = curstate
+    bestcost,dummy1,dummy2 = best_costs.get(state)   
+    retpath: List[str] = []
+    while(1):       
+        cost,rname,fromstate  = best_costs.get(state)      
+        if(rname == "") :           
+            break;
+        retpath.insert(0,rname)       
+        state = fromstate
+      
+    print("plan_dijkstra visited=",visit_count,"best_cost=",bestcost,"retpath=",retpath)
+    return (visit_count, bestcost, retpath)       
+
+def plan_width(initial: State, goal: State, WMax: int) -> Tuple[int, int, Optional[List[str]]]:
+    all_propositions = recipe_propositions | state_propositions(initial) | state_propositions(goal)
+    all_combinations: List[FrozenSet[Proposition]] = []
+    for W in range(1, WMax + 1):            
+        start_time = time.time()       
+        visit_count = 0
+        # Calculate all combinations of propositions at size W and add to all_combinations
+        all_combinations += [frozenset(props) for props in itertools.combinations(all_propositions, W)]
+        # Sanity check that this is 6279 for W=3, for example
+        print("W=",W,"Combination count=",len(all_combinations))
+        # Track, for each combination (by index), whether we have seen this combination before (0 for no, >0 for yes)
+        
+        seen_combinations: Set[FrozenSet[Proposition]] = set()
+                
+        open_list: List[Tuple[int, State,str]] = [(0, initial, "")]         
+        best_costs: Dict[State, Tuple[int, str,State]] = {initial:(0, "",initial)}       
+        found = False
+        while open_list:
+            # Dijkstra's search uses the priority queue data structure
+            cost, curstate , gotfrom  = heapq.heappop(open_list)       
+            
+            #print("cost=",cost,"curstate=",curstate,"from=",gotfrom)
+           
+            if(curstate >= goal):
+                x,rname,z = best_costs.get(curstate)
+                print ("found it!! Number of visited nodes=" ,visit_count , "state=",curstate,"goal=",goal,"by=",rname) 
+                found = True;
+                break;
+                    
+            visit_count += 1
+                                        
+            for candidaterecipename in recipes:
+                if (preconditions_satisfied(curstate,recipes[candidaterecipename])) == False:
+                    continue 
+                
+                candidatestate = apply_effects(curstate,recipes[candidaterecipename])
+                IsAddingValue = see_state(candidatestate,all_combinations,seen_combinations) 
+                if(IsAddingValue == False):
+                    continue
+               
+                                
+                if(candidatestate in  best_costs):
+                    #print("found=" , neighbor)     
+                    candidatestate_cost,rname,fromstate = best_costs.get(candidatestate)
+                    if(candidatestate_cost > cost + recipes[candidaterecipename].cost) :        
```

Comment:
I think I mentioned it before but might help to save the `cost + recipes[candidaterecipename].cost` as a local variable. And here it helps a little more since then you do not have to access recipes[] each use, which would save some time.